[16.0][IMP] mis_builder: introduce annotation in reports

AnizR opened the pull request:
This PR introduces 'annotation' in mis_builder reports.
An annotation is a comment or a remark that a user can write on a cell.

It tries to solve issue: #159 (opened 6 yeas ago!)

# Description
The goal is to write a remark on a cell (independently of the pivot date).
These annotations are visible directly within the view but also at export in .pdf and in .xlsx.

## Overview
### UI in mis_builder_widget
On cells, we have now a drop down menu to drill down or annotate:
![image](https://github.com/user-attachments/assets/389e27bf-d7ee-48d6-812a-d677b0bb6795)
It opens a window to write an annotation:
![image](https://github.com/user-attachments/assets/71dc09e0-0799-4f26-a502-a7478f060e2f)
**Before refreshing**, annotations are pushed to the db, added in the 'footnotes' and if we try to 're-annotate', the annotation is displayed but the sequence is not yet set.
**After refreshing**, annotations are numbered. I did this to avoid the user to force to refresh his report each time he writes an annotation.
![image](https://github.com/user-attachments/assets/b33563a3-3abc-43d9-9385-a085307ebbf2)

### Print .pdf

The annotation is also present in the .pdf:
![image](https://github.com/user-attachments/assets/4a308eda-04ad-4c30-a70f-2f5e6750f438)


### Export .xls

In the .xls, the annotation are added as 'comment' on cells:
![image](https://github.com/user-attachments/assets/62e333e9-78f1-40e4-915e-a187f239cfc3)

hitrosol commented:
Long wait feature finally come. 

I tried on runboat using Chrome, Edge and Safari. The drop down menu is behind the other row.

![Screen Shot 2025-04-03 at 11 19 37 PM](https://github.com/user-attachments/assets/30d747df-49e8-4a94-8bba-cab1eac7436f)


It looks okay on Firefox, but no sequence number is added in the footnotes. If I refresh the browser the notes is gone.
![Screen Shot 2025-04-03 at 11 21 59 PM](https://github.com/user-attachments/assets/91510370-2264-448b-b119-7c3f3e49b601)

Thanks

sbidoul commented:
> Long wait feature finally come.

It was just waiting for someone to fund it.

AnizR commented:
> Long wait feature finally come.
> 
> I tried on runboat using Chrome, Edge and Safari. The drop down menu is behind the other row.
> 
> ![Screen Shot 2025-04-03 at 11 19 37 PM](https://private-user-images.githubusercontent.com/5949936/430049093-30d747df-49e8-4a94-8bba-cab1eac7436f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM3MDczNTAsIm5iZiI6MTc0MzcwNzA1MCwicGF0aCI6Ii81OTQ5OTM2LzQzMDA0OTA5My0zMGQ3NDdkZi00OWU4LTRhOTQtOGJiYS1jYWIxZWFjNzQzNmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDNUMTkwNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTliZjE0MDUwZmI1Y2E5NjZlMWFjODkxM2JhYWI0MjcwMmFlOTBjMGU4MzcwYjRjMjJlZjUzNmZmMjg5NGFkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.8IK8xfniuxpkQ9UHeUPFli0IPEw2ysD4RSRtj5btM-w)
> 
> It looks okay on Firefox, but no sequence number is added in the footnotes. If I refresh the browser the notes is gone. ![Screen Shot 2025-04-03 at 11 21 59 PM](https://private-user-images.githubusercontent.com/5949936/430049733-91510370-2264-448b-b119-7c3f3e49b601.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM3MDczNTAsIm5iZiI6MTc0MzcwNzA1MCwicGF0aCI6Ii81OTQ5OTM2LzQzMDA0OTczMy05MTUxMDM3MC0yMjY0LTQ0OGItYjExOS03YzNmM2U0OWI2MDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDNUMTkwNDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjA0NmZjM2M5NTI3NjI3YTRhY2E1YTcwNzIxMDBjM2ViODM2MThhNzk2MDhjNGVhNTAyYmMxZWFhN2E4NjE3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yPAJe6lGG2vjtsDvrRaeWsmCSt6a7glZgUaUnaq0a0k)
> 
> Thanks

Thanks for your feedback.
I am using firefox but I'll try to figure out why it doesn't work with other browsers.
Regarding the annotations, I opened your instance and indeed, it doesn't work with your column '3 M Budget' but it works with others. I'll fix it:
![image](https://github.com/user-attachments/assets/af930869-b829-4271-b984-bea4da71c781)

AnizR commented:
> Long wait feature finally come.
> 
> I tried on runboat using Chrome, Edge and Safari. The drop down menu is behind the other row.
> 
> ![Screen Shot 2025-04-03 at 11 19 37 PM](https://private-user-images.githubusercontent.com/5949936/430049093-30d747df-49e8-4a94-8bba-cab1eac7436f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDA4MzYsIm5iZiI6MTc0NDA0MDUzNiwicGF0aCI6Ii81OTQ5OTM2LzQzMDA0OTA5My0zMGQ3NDdkZi00OWU4LTRhOTQtOGJiYS1jYWIxZWFjNzQzNmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDdUMTU0MjE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTlkOTZlMmUzZTQ2ODBiOWQzZTRiZmY4N2NhZDM2MzhlOTA1NzE3ZTE0MDlmNjNjY2I4NjBhNmE0M2Y4N2U1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EYStNdBJ6iQseIekN-b_1-NJeBw02z978qryC-VPojw)
> 
> It looks okay on Firefox, but no sequence number is added in the footnotes. If I refresh the browser the notes is gone. ![Screen Shot 2025-04-03 at 11 21 59 PM](https://private-user-images.githubusercontent.com/5949936/430049733-91510370-2264-448b-b119-7c3f3e49b601.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwNDA4MzYsIm5iZiI6MTc0NDA0MDUzNiwicGF0aCI6Ii81OTQ5OTM2LzQzMDA0OTczMy05MTUxMDM3MC0yMjY0LTQ0OGItYjExOS03YzNmM2U0OWI2MDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDdUMTU0MjE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzQzZTNjZTgyNjMxNDhiZTdkM2VjOWYzMzE3N2QyZTdhNDBkMzhhZTc1ZDhiNzBhOWUyNzg2ODFhNGE2ZWY0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ECWXL-fwTn2Fnm9WIRzz26LLXD2CloC8OCppcXXiWJ8)
> 
> Thanks

It should be fixed now.
Thanks for your feedback!

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,56 @@
+    note = fields.Char()
+
+    @api.model
+    def edit_annotation(self, period_id, kpi_id, subkpi_id, note):
```
```suggestion
    def set_annotation(self, period_id, kpi_id, subkpi_id, note):
```

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -801,6 +801,17 @@ def _add_column_move_lines(self, aep, kpi_matrix, period, label, description):
+        note_by_subkpi_id = {
+            (note.kpi_id.id, note.subkpi_id.id): note.note for note in annotations
+        }
+        locals_dict = {"notes": note_by_subkpi_id}
```
I need to look closer, but locals_dict is used as the evaluation context for expressions. I don't think it is useful to add notes in that context

AnizR replied:
Indeed, I refactored a bit to avoid using the locals_dict.
Instead, I directly added the notes when adding a column which enables adding annotation on budget and comparison columns.

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,56 @@
+    )
+    subkpi_id = fields.Many2one(
+        comodel_name="mis.report.subkpi",
+    )
```
Add a unique constraint on these 3 fields?

sbidoul commented on mis_builder/security/ir.model.access.csv:
```diff
@@ -20,3 +20,4 @@ access_mis_report_subreport,access_mis_report_subreport,model_mis_report_subrepo
 manage_mis_report_style,access_mis_report_style,model_mis_report_style,account.group_account_manager,1,1,1,1
 access_mis_report_style,access_mis_report_style,model_mis_report_style,base.group_user,1,0,0,0
 access_add_to_dashboard_wizard,access_add_to_dashboard_wizard,model_add_mis_report_instance_dashboard_wizard,base.group_user,1,1,1,0
+access_mis_report_annotation, access_mis_report_annotation,model_mis_report_instance_annotation,base.group_user,1,1,1,1
```
That is too broad? Annotation can possibly contain sensitive data so I would not have that readable by all employees by default.

AnizR replied:
To be as flexible as possible, I introduced 2 new groups to edit and read annotations

sbidoul reviewed (commented):
Thanks @AnizR. Here is a first round of comment after a quick look.

SAnnabelle reviewed (changes requested):
It works well! 

Would it be possible to make the annotations company-dependent? This would allow each company to have its own comments if the report is shared between several companies.

AnizR commented:
> It works well!
> 
> Would it be possible to make the annotations company-dependent? This would allow each company to have its own comments if the report is shared between several companies.

Thanks for your feedback!

Annotations are now linked to a company.
This company is the 'main' company of the user when creating the annotation.

sbidoul commented:
> Annotations are now linked to a company.
> This company is the 'main' company of the user when creating the annotation.

Is that not potentially surprising for the user?

For instance, what if it is a multi-company consolidation report?

sbidoul commented:
I wonder to what extent the annotation should be linked to (some of) the filter values of the report (effective companies, pivot date, ...).

AnizR commented:
> > Annotations are now linked to a company.
> > This company is the 'main' company of the user when creating the annotation.
> 
> Is that not potentially surprising for the user?
> 
> For instance, what if it is a multi-company consolidation report?

I figured that the 'final user' would always use the same 'main company' when watching his report. Therefore, in my case, it makes sense to attach annotations to the 'main company'. What would you prefer? Storing every companies selected when watching the report?


> I wonder to what extent the annotation should be linked to (some of) the filter values of the report (effective companies, pivot date, ...).

I thought about adding the pivot date but realized that in odoo enterprise, they also have an 'annotation' feature but they don't store the pivot date. The annotation is linked to the 'cell'.

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,71 @@
+    )
+    subkpi_id = fields.Many2one(
+        comodel_name="mis.report.subkpi",
+    )
```
`ondelete="cascade"` on period_id, kpi_id, subkpi_id ?

Some `required=True`?

Also add a composite index on `period_id, kpi_id, subkpi_id` to anticipate when many old notes accumulate?

AnizR replied:
done :+1:

AnizR commented:
> I wonder to what extent the annotation should be linked to (some of) the filter values of the report (effective companies, pivot date, ...).

I changed a bit my code:

- I removed the field 'company_id' on annotations
- I introduced a 'annotation_context' that represents the context in which an annotation has been added and in which we should display an annotation.

For now, I only use the 'query_company_ids' from the mis report instance.

So, let's say that we have a report that can be used on company A and B. I am on company A and I add a note. This note will be visible only if I display the report from company A (without company B). If I add a note while having company A and B selected, this note will be displayed on the report only if I have company A and B selected.

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,106 @@
+                    "kpi_id": kpi_id,
+                    "subkpi_id": subkpi_id,
+                    "note": note,
+                    "annotation_context": self._get_annotation_context(instance_id),
```
```suggestion
                    "annotation_context": json.dumps(instance_id._get_annotation_context()),
```

It's simpler.

But is the `json.dumps` necessary? I'd think we can provide the dict to the json field.

AnizR replied:
Indeed the `json.dumps` can be safely removed

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -940,3 +963,15 @@ def _get_drilldown_action_name(self, arg):
             return f"{kpi.description} - {period.display_name}"
+
+    def _get_annotation_context(self):
+        """
+        Return context of annotation linked to this instance
+        """
```
```suggestion
        """Return the context used to filter annotation linked to this instance."""
```

Also, nit: https://peps.python.org/pep-0257/

sbidoul commented on mis_builder/security/res_groups.xml:
```diff
@@ -0,0 +1,13 @@
+            eval="[Command.link(ref('mis_builder.group_read_annotation'))]"
+        />
+    </record>
+</odoo>
```
Automatically assign these new groups to the admin user? I think there is a pattern for that throughout Odoo.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.xml:
```diff
@@ -85,21 +85,49 @@
+                                                t-if="cell.drilldown_arg"
+                                                role="menuitem"
+                                            >
+                                                Explore
```
Drill down instead of Explore ? not sure.

But can we keep the immediate drill down when we click on the value, and have the menu appear only when we click on the little down arrow ?

AnizR replied:
Ok, I'll remove the 'Explore' in the menu and set back the immediate drill down when clicking on cell's value

sbidoul commented on mis_builder/security/res_groups.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record model="res.groups" id="group_read_annotation">
+        <field name="name">Mis report: read annotation</field>
```
```suggestion
        <field name="name">MIS Report: view annotations</field>
```

sbidoul commented on mis_builder/security/res_groups.xml:
```diff
@@ -0,0 +1,13 @@
+        <field name="name">Mis report: read annotation</field>
+    </record>
+    <record model="res.groups" id="group_edit_annotation">
+        <field name="name">Mis report: edit annotation</field>
```
```suggestion
        <field name="name">MIS report: add annotations</field>
```

sbidoul reviewed (commented):
Here are a few remarks about ergonomy:

- Now the cells where there is no drill down are underlined on hover. It is important that this underline appears only as visual cue that a drilldown is possible, as before.
- After adding an annotation I had to refresh to see it. Is that intended?
- Removing an annotation leaves a weird display, with the little call out still present until I refresh

![image](https://github.com/user-attachments/assets/ce52d7be-55dd-439f-9a18-a3a6e0665799)

I think it's important that the annotation changes appear correctly immediately, and without requiring a full recompute.

Tested with firefox on the demo report on runboat.

AnizR commented:
> Here are a few remarks about ergonomy:
> 
>     * Now the cells where there is no drill down are underlined on hover. It is important that this underline appears only as visual cue that a drilldown is possible, as before.
> 
>     * After adding an annotation I had to refresh to see it. Is that intended?
> 
>     * Removing an annotation leaves a weird display, with the little call out still present until I refresh
> 
> 
> ![image](https://private-user-images.githubusercontent.com/692075/432699771-ce52d7be-55dd-439f-9a18-a3a6e0665799.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3MDczNDIsIm5iZiI6MTc0NDcwNzA0MiwicGF0aCI6Ii82OTIwNzUvNDMyNjk5NzcxLWNlNTJkN2JlLTU1ZGQtNDM5Zi05YTE4LWEzYTZlMDY2NTc5OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNVQwODUwNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZjc3OWYyNDU4YjU3ZTQzNmZhNTcwYjZkZTQ4MjQwYjRiMWU1OTcxNWVhMTJmZGI5ZDFlNTEwM2JkNjU3MDhiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8nBkabnkKL9T5wmikApddM1z8C4tf0MSWM2FLEduMIA)
> 
> I think it's important that the annotation changes appear correctly immediately, and without requiring a full recompute.
> 
> Tested with chrome on the demo report on runboat.

Thanks for your feedback.
I have done some changes, those ergonomic issues are fixed.
Now, the loading of notes can be triggered independently from the computation of the matrix of a report!

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -524,16 +523,47 @@ def as_dict(self):
                             cell.style_props, no_indent=True
                         ),
+                        # notes can not be added on 'details by account' lines
+                        "can_be_annotated": not bool(cell.row.account_id),
```
```suggestion
                        "can_be_annotated": not cell.row.account_id,
```

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -524,16 +523,47 @@ def as_dict(self):
+    # logic to convert semantic coordinates (period, kpi, subkpi)
+    # to visual coordinates (cell id)
+    @classmethod
+    def _make_row_id(cls, kpi_id, account_id) -> str:
```
```suggestion
    def _make_row_id(cls, kpi_id: int, account_id: int | None) -> str:
```

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -524,16 +523,47 @@ def as_dict(self):
         return {"header": header, "body": body}
+
+    # logic to convert semantic coordinates (period, kpi, subkpi)
+    # to visual coordinates (cell id)
```
```suggestion
    # Logic to convert semantic coordinates (period, kpi, subkpi)
    # to visual coordinates (cell id) and back. The rendering logic must no know
    # about semantic concepts such as periods and kpis. Having these well identified
    # methods allow us to easily spot where the conversion between the rendering and
    # semantic domain occur.

```

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -524,16 +523,47 @@ def as_dict(self):
+        return f"{kpi_id}#{account_id or ''}#{period_id}#{subkpi_id or ''}"
+
+    @classmethod
+    def _pack_cell_id(cls, cell):
```
```suggestion
    def _pack_cell_id(cls, cell: KpiMatrixCell) -> str:
```

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -524,16 +523,47 @@ def as_dict(self):
+        )
+
+    @classmethod
+    def _unpack_cell_id(cls, cell_id: str):
```
```suggestion
    def _unpack_cell_id(cls, cell_id: str) -> Tuple[int, int | None, int, int | None:
         """Convert cell id to kpi_id, account_id, period_id, subkpi_id."""
```

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -877,7 +878,35 @@ def _compute_matrix(self):
+                annotation.subkpi_id and annotation.subkpi_id.id,
+            ): {"note_text": annotation.note, "note_sequence": sequence}
+            for sequence, annotation in enumerate(annotations, 1)
+        }
```
Computing the sequence without sorting here may lead to surprising numbering. I see two possibilities
1. assign sequence numbers in the rendering phase (i.e. in de widget, and pdf - in xls we don't need it)
2. or sort the annotations here, probably by (period.sequence, subkpi.sequence, kpi.sequence), altough this sorting algo would gain quite a bit of knowledge of the rendering algorithm

Granted, (2) my be hard to do in a qweb template.

AnizR replied:
I prefer the second option with a small tweak, I'll sort the annotation by (kpi.sequence,period.sequence,subkpi.sequence) to follow the 'reading order' (left to right and up to bottom)

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -877,7 +878,35 @@ def _compute_matrix(self):
+        ret["notes"] = self.get_notes()
+        return ret
+
+    def get_notes(self) -> dict:
```
```suggestion
    def get_notes_by_cell_id(self) -> dict:
        """Return annotations by cell id."""
```

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -940,3 +969,13 @@ def _get_drilldown_action_name(self, arg):
+        """Return the context used to filter annotation linked to this instance."""
+        self.ensure_one()
+        annotation_context = {}
+        if companies := self.query_company_ids.ids:
```
```suggestion
        if query_company_ids := self.query_company_ids.ids:
```

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,105 @@
+        ondelete="cascade",
+    )
+    note = fields.Char()
+
```
```suggestion
```

sbidoul commented on mis_builder/models/kpimatrix.py:
```diff
@@ -144,6 +141,8 @@ def __init__(
         self.drilldown_arg = drilldown_arg
         self.val_type = val_type
 
```
```suggestion
```

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,105 @@
+        (
+            "annotation_unique",
+            "unique(period_id, kpi_id,subkpi_id)",
+            "Only one annotation per cell(defined using period,kpi,sub kpi)",
```
```suggestion
            "Only one annotation per cell is allowed (defined using period, kpi, sub kpi)",
```

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,105 @@
+        )
+
+    @api.model
+    def get_first_matching_annotation(self, cell_id, instance_id):
```
```suggestion
    def _get_first_matching_annotation(self, cell_id, instance_id):
```

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -877,7 +878,35 @@ def _compute_matrix(self):
+                False,
+                annotation.period_id.id,
+                annotation.subkpi_id and annotation.subkpi_id.id,
+            ): {"note_text": annotation.note, "note_sequence": sequence}
```
```suggestion
            ): {"text": annotation.note, "sequence": sequence}
```

In the contexts where we use that we know we are talking about notes, so this would like make things lighter.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -65,3 +65,14 @@
+
+.footnote-mis-builder {
+    font-size: 80%;
+    color: #fd0000;
```
I think there are color variables we can and should use.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.esm.js:
```diff
@@ -138,6 +142,21 @@ export class MisReportWidget extends Component {
+        );
+        this.state.can_read_annotation = await this.user.hasGroup(
+            "mis_builder.group_read_annotation"
+        );
```
Let's add fields on mis.report.instance so the logic to decide if annotations can be read or written is easier to override.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.xml:
```diff
@@ -85,21 +86,52 @@
+                                            role="menu"
+                                        >
+                                            <a
+                                                t-if="state.can_edit_annotation and cell.can_be_annotated"
```
Always show the Annotate menu, but gray it out / disable it if annotations are not supported.

Actually, if `state.can_edit_annotation` is False, we can hide the dropdown triangle completely. And if `cell.can_be_annotated` is False, show it with a diabled Annotate menu entry.

sbidoul reviewed (approved):
The code is much easier to read, thanks!

AnizR commented:
I have done some changes and improvements that were asked.

I've done multiple commits in order to make the review easier. Once everything will be "approved", I'll squash all my commits.

sbidoul commented on mis_builder/models/mis_report_instance.py:
```diff
@@ -877,7 +884,44 @@ def _compute_matrix(self):
+
+    def get_notes_by_cell_id(self) -> dict:
+        self.ensure_one()
+        if not self.env.user.has_group("mis_builder.group_read_annotation"):
```
```suggestion
        if not self.user_can_read_annotation:
```

sbidoul commented on mis_builder/models/mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,104 @@
+    def remove_annotation(self, cell_id, instance_id):
+        annotation = self._get_first_matching_annotation(cell_id, instance_id)
+        if annotation:
+            annotation.unlink()
```
Should we test `instance.user_can_edit_annotation` here?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -65,3 +65,18 @@
+    bottom: 1ex;
+}
+
+.annotation_disabled {
```
not used?

sbidoul commented on mis_builder/tests/test_mis_report_instance_annotation.py:
```diff
@@ -0,0 +1,165 @@
+                            description="subkpi 2, report 2",
+                            sequence=2,
+                        ),
+                    ),
```
Use `Command.create`?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.esm.js:
```diff
@@ -140,6 +148,15 @@ export class MisReportWidget extends Component {
+    async refresh_annotation() {
+        this.state.mis_report_data.notes = await this.orm.call(
+            "mis.report.instance",
+            "get_notes",
```
```suggestion
            "get_notes_by_cell_id",
```

sbidoul commented on mis_builder/report/mis_report_instance_qweb.xml:
```diff
@@ -97,12 +98,30 @@
                                                 />
+                                                <span class="footnote-mis-builder">
+                                                    <t
+                                                        t-out="notes.get(cell.cell_id,{}).get('sequence','')"
```
I got a `AttributeError: 'NoneType' object has no attribute 'cell_id'` when generating the pdf with the default demo report on runboat, probably here.

AnizR replied:
In deed, I'll check that, thanks!

sbidoul reviewed (commented):
A few more details.

If disabled menu items don't work I really prefer to have the little triangle with an empty menu so figures stay aligned. Can yo also put a little space between the cell content and the arrow?

![image](https://github.com/user-attachments/assets/beda0ce2-d43f-4a6e-bcb3-d36d6898fc05)


Otherwise I think its almost done. Thanks a lot!

AnizR commented:
> A few more details.
> 
> If disabled menu items don't work I really prefer to have the little triangle with an empty menu so figures stay aligned. Can yo also put a little space between the cell content and the arrow?
> 
> ![image](https://private-user-images.githubusercontent.com/692075/434322235-beda0ce2-d43f-4a6e-bcb3-d36d6898fc05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4Nzc5MDQsIm5iZiI6MTc0NDg3NzYwNCwicGF0aCI6Ii82OTIwNzUvNDM0MzIyMjM1LWJlZGEwY2UyLWQ0M2YtNGE2ZS1iY2IzLWQzNmQ2ODk4ZmMwNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QwODEzMjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ODFkOGM0ZDEyYzM2ZjgzMTE3OTA1ZDNhYjUwOWM5ZDNiNzNkNjFkNWQ0OWM2N2FmNGViZWEzMzcyYTZjY2RlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tt0yS38HE3VxIwV0xFPrHznpIZeoLuLLKOnjQa5fszo)
> 
> Otherwise I think its almost done. Thanks a lot!

Thanks again for your review.
Here is how it looks now with some padding and the drop-down button greyed out:
![image](https://github.com/user-attachments/assets/ddc876c7-5b43-479c-a615-2f60c8b8b69b)

sbidoul commented:
There is still an alignment problem: 
![image](https://github.com/user-attachments/assets/11dfe2fd-c860-43c5-abeb-29967e71ab31)

sbidoul commented:
Also, can we tweak the pdf a little bit?

- In cells, make a reserved space for the footnoote callout so figures remain aligned? Add this reserved space in all cells only if there is at least one note?
- Remove the bullets in the bottom footnotes list (in books or documents we usually see the footnote number, it's enough and nicer).
- Add a some padding between the bottom footnotes list and the table.

![image](https://github.com/user-attachments/assets/b2d82ea8-9592-4280-bcb3-968cc78e47ee)

sbidoul commented:
> In cells, make a reserved space for the footnoote callout so figures remain aligned? Add this reserved space in all cells only if there is at least one note?

In the widget too?

sbidoul commented on mis_builder/static/src/components/mis_report_widget.esm.js:
```diff
@@ -170,6 +187,69 @@ export class MisReportWidget extends Component {
+    }
+
+    async remove_annotation(event) {
+        const cell_id = $(event.target).data("cell-id");
```
Do we really need these $ ? In v18 there is no jquery anymore, so let's not add more dependency on it, that will make migration harder.

sbidoul commented on mis_builder/static/src/components/mis_report_widget.esm.js:
```diff
@@ -170,6 +187,69 @@ export class MisReportWidget extends Component {
+                })
+            ),
+            buttons: buttons,
+        }).open();
```
A little detail if not hard to do: put the focus on the content field when opening the dialog box.

AnizR replied:
I tried but didn't manage do it easily

AnizR commented:
> There is still an alignment problem: ![image](https://private-user-images.githubusercontent.com/692075/434751288-11dfe2fd-c860-43c5-abeb-29967e71ab31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4OTgzMzAsIm5iZiI6MTc0NDg5ODAzMCwicGF0aCI6Ii82OTIwNzUvNDM0NzUxMjg4LTExZGZlMmZkLWM4NjAtNDNjNS1hYmViLTI5OTY3ZTcxYWIzMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QxMzUzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZDQ4MzQyOTA3MTRlYTJkZGMyOWY4MWFjY2FiMjZiMGEwYWIyMzQ5MzdlMjI0YzgyYzZiMWE5NTA1ODBiNzEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.B8kdh7GqQNaj6Hb5znrUbwIRYdfW6swOCKjUx7JDEn4)

Finally :+1: 
![image](https://github.com/user-attachments/assets/52d64a7f-e7c6-46de-b82a-0fecd1b6b345)

SAnnabelle commented:
1. The structure of the note with line breaks is not retained in footnotes.

![image](https://github.com/user-attachments/assets/df224793-2681-464b-aab6-76bcc5227925)

![image](https://github.com/user-attachments/assets/ed656c42-8062-402d-b901-5c211e4b636b)

2. Is it possible that when we click on the footnote number, we are redirected to the footnote in the report?

sbidoul commented:
Rich text annotations might be a can of worms :) And it's not really a footnote anymore (if we compare to books or documents). Is that an important use case?

sbidoul commented:
> Is it possible that when we click on the footnote number, we are redirected to the footnote in the report?

Which makes me think. In the widget is it not better to display footnotes on hover? In large reports where the footnotes be far below the view it would make them easier to view.

AnizR commented:
> > Is it possible that when we click on the footnote number, we are redirected to the footnote in the report?
> 
> Which makes me think. In the widget is it not better to display footnotes on hover? In large reports where the footnotes be far below the view it would make them easier to view.

I'll add it on hover but it is possible to directly see the annotation by opening the wizard to edit the annotation

sbidoul commented:
> It is possible to directly see the annotation by opening the wizard to edit the annotation

Except if you are a readonly user?

AnizR commented:
> Rich text annotations might be a can of worms :) And it's not really a footnote anymore (if we compare to books or documents). Is that an important use case?

What I can do is allowing multi line note and display as:
![image](https://github.com/user-attachments/assets/83ebe3c0-afdf-4847-bdf9-a93fdb4c0062)



> > It is possible to directly see the annotation by opening the wizard to edit the annotation
> 
> Except if you are a readonly user?

In deed!

AnizR commented:
I've added:
1. Displaying note when hovering the sequence of annotation
2. Clicking on annotations triggers a scroll down to the section of note
3. In the footnotes, when clicking on the number of a note, it will scroll up to the corresponding note
4. Multi line notes are displayed in multi line instead of forcing it into one line

sbidoul commented on mis_builder/static/src/components/mis_report_widget.css:
```diff
@@ -80,8 +80,14 @@
 }
 
-.footnote_mis_builder_list {
+.footnote_mis_builder_table {
```
I had not noticed this, but let's prefix all css classes with `.oe_mis_builder` ?

SAnnabelle reviewed (approved):
Functionally, it's very good. 

- Access rights work
- The multi-company context is taken into account
- Notes are also included in the pdf file and excel export
- Several things have been put in place to make it easier for users to create and view notes

AnizR commented:
> The 'drill down' seems to be broken, I'll fix it.

Fixed!

sbidoul commented:
If there are no additional comments I'll merge next week.

sbidoul commented:
/ocabot merge minor

hbrunn commented:
note this breaks with python 3.9

sbidoul commented:
Ah, yes, we don't test nor officially support Python 3.9 in OCA 16.0.
What's breaking exactly?

hbrunn commented:
the type hints, `sometype | someothertype` was introduced in 3.10

sbidoul commented:
> the type hints, `sometype | someothertype` was introduced in 3.10

Right. We'll fix that.

hbrunn commented:
I'll be happy to PR removing the `| None` parts, the rest happily works with py3.9. just wanted to be sure you actually care about 3.9
